aidial_adapter_bedrock/llm/message.py

```diff
@@ -1,48 +1,92 @@
+    content: str
+
+
+class AIToolCallMessage(BaseModel):
+    calls: List[ToolCall]
+
+
+class AIFunctionCallMessage(BaseModel):
+    call: FunctionCall
 
 
-class SystemMessage(BaseMessage):
-    @property
-    def type(self) -> str:
-        return "system"
+BaseMessage = Union[SystemMessage, HumanRegularMessage, AIRegularMessage]
+ToolMessage = Union[
+    HumanToolResultMessage,
+    HumanFunctionResultMessage,
+    AIToolCallMessage,
+    AIFunctionCallMessage,
+]
 
 
-class HumanMessage(BaseMessage):
-    @property
-    def type(self) -> str:
-        return "human"
+def _parse_assistant_message(
+    content: Optional[str],
+    function_call: Optional[FunctionCall],
+    tool_calls: Optional[List[ToolCall]],
+) -> BaseMessage | ToolMessage:
+    if content is not None and function_call is None and tool_calls is None:
+        return AIRegularMessage(content=content)
 
+    if content is None and function_call is not None and tool_calls is None:
+        return AIFunctionCallMessage(call=function_call)
 
-class AIMessage(BaseMessage):
-    @property
-    def type(self) -> str:
-        return "ai"
+    if content is None and function_call is None and tool_calls is not None:
+        return AIToolCallMessage(calls=tool_calls)
 
+    raise ValidationError(
+        "Assistant message must have one and only one of the following fields not-none: "
+        f"content (is none: {content is None}), "
+        f"function_call (is none: {function_call is None}), "
+        f"tool_calls (is none: {tool_calls is None})"
+    )
 
-def parse_message(msg: Message) -> BaseMessage:
-    if msg.content is None:
-        raise ValidationError("Message content must be present")
 
-    match msg.role:
-        case Role.SYSTEM:
-            return SystemMessage(content=msg.content)
-        case Role.USER:
-            return HumanMessage(content=msg.content)
-        case Role.ASSISTANT:
-            return AIMessage(content=msg.content)
-        case Role.FUNCTION:
-            raise ValidationError("Function calls are not supported")
+def parse_message(msg: Message) -> BaseMessage | ToolMessage:
+    match msg:
+        case Message(role=Role.SYSTEM, content=content) if content is not None:
+            return SystemMessage(content=content)
+        case Message(role=Role.USER, content=content) if content is not None:
+            return HumanRegularMessage(content=content)
+        case Message(
+            role=Role.ASSISTANT,
+            content=content,
+            function_call=function_call,
+            tool_calls=tool_calls,
+        ):
+            return _parse_assistant_message(content, function_call, tool_calls)
+        case Message(
+            role=Role.FUNCTION, name=name, content=content
+        ) if content is not None and name is not None:
+            return HumanFunctionResultMessage(name=name, content=content)
+        case Message(
+            role=Role.TOOL, tool_call_id=id, content=content
```

Comment:
Will it work if name is set? Here in the cookbook examples the name is set to function name: https://cookbook.openai.com/examples/how_to_call_functions_with_chat_models